plugins/out_cloudwatch_logs/cloudwatch_api.c

```diff
@@ -991,6 +992,17 @@ int put_log_events(struct flb_cloudwatch *ctx, struct cw_flush *buf,
     flb_sds_t error;
     int num_headers = 1;
 
+    buf->put_events_calls++;
+
+    if (buf->put_events_calls >= 4) {
+        /*
+         * In normal execution, even under high throughput, 4+ calls per flush
+         * should be extremely rare. This is needed for edge cases basically.
+         */
+        flb_plg_debug(ctx->ins, "Too many calls this flush, sleeping for 250 ms");
+        usleep(250000);
```

Comment:
If this function ```put_log_events``` is under a flush() invocation/coroutine, pls use ```flb_time_sleep(a, b)``` instead of ```usleep()```. The proposed function does an async sleep (non-blocking)

Reply:
Uhhh... I'm not sure this will work actually... this plugin is meant to be entirely synchronous- the code was written under that assumption. This is because the CW Logs API requires Puts to a stream to be synchronous. (Long term, if we introduce some more complicated concurrency mechanisms in Fluent Bit, we can make it concurrent). 

A CW Logs Put request is 1 MB, and there's a few buffers the plugin needs each flush in order to process the logs and construct the payload. Since I knew the plugin was going to be synchronous, I just allocate these buffers in init one time and re-use them in every flush. I did some benchmarking and this made it way faster; allocating memory in each flush slowed things down. Anyway- this means that any introduction of concurrency could break the code.

I think I prefer to keep the code as is, because as stated in the comment, this is an edge/rare case. The synchronous sleep here should not be triggered very often in the real world.